src/test/java/com/google/cloud/teleport/io/WindowedFilenamePolicyTest.java

```diff
@@ -214,18 +218,126 @@ public void testWithDynamicDirectory() throws IOException {
+
+    assertThat(filename, is(notNullValue()));
+    assertThat(
+        filename.getCurrentDirectory().toString().endsWith("2017/1/8/10:56/"), is(equalTo(true)));
+    assertThat(filename.getFilename(), is(equalTo("output-001-of-001")));
+  }
+
+  /**
+   * Tests that windowedFilename() produces the correct directory when using a custom DateTime
+   * String.
+   */
+  @Test
+  public void testCustomDynamicDirectoryString() {
+
+    ResourceId outputDirectory =
+        getBaseTempDirectory().resolve("y/M/D/H:m", StandardResolveOptions.RESOLVE_DIRECTORY);
+    IntervalWindow window = mock(IntervalWindow.class);
+    PaneInfo paneInfo = PaneInfo.createPane(false, true, Timing.ON_TIME, 0, 0);
+
+    Instant windowBegin = new DateTime(2017, 1, 8, 10, 55, 0).toInstant();
+    Instant windowEnd = new DateTime(2017, 1, 8, 10, 56, 0).toInstant();
+    when(window.maxTimestamp()).thenReturn(windowEnd);
+    when(window.start()).thenReturn(windowBegin);
+    when(window.end()).thenReturn(windowEnd);
+
+    WindowedFilenamePolicy policy =
+        WindowedFilenamePolicy.writeWindowedFiles()
+            .withOutputDirectory(outputDirectory.toString())
+            .withOutputFilenamePrefix("output")
+            .withShardTemplate("-SSS-of-NNN")
+            .withSuffix("")
+            .withYearPattern("y")
+            .withMonthPattern("M")
+            .withDayPattern("D")
+            .withHourPattern("H")
+            .withMinutePattern("m");
+
+    ResourceId filename =
+        policy.windowedFilename(1, 1, window, paneInfo, new TestOutputFileHints());
+
+    assertThat(filename, is(notNullValue()));
+    assertThat(
+        filename.getCurrentDirectory().toString().endsWith("2017/1/8/10:56/"), is(equalTo(true)));
+    assertThat(filename.getFilename(), is(equalTo("output-001-of-001")));
+  }
+
+  /**
+   * Tests that windowedFilename() throws a {@link RuntimeException} if an invalid DateTime pattern
+   * is provided.
+   */
+  @Test
+  public void testInvalidDynamicDirectory() {
+    exception.expect(RuntimeException.class);
+    final String invalidMinutePattern = "i/m";
+
+    ResourceId outputDirectory =
+        getBaseTempDirectory()
+            .resolve(invalidMinutePattern, StandardResolveOptions.RESOLVE_DIRECTORY);
+    IntervalWindow window = mock(IntervalWindow.class);
+    PaneInfo paneInfo = PaneInfo.createPane(false, true, Timing.ON_TIME, 0, 0);
+
+    WindowedFilenamePolicy policy =
+        WindowedFilenamePolicy.writeWindowedFiles()
+            .withOutputDirectory(StaticValueProvider.of(outputDirectory.toString()))
+            .withOutputFilenamePrefix(StaticValueProvider.of("output"))
+            .withShardTemplate(StaticValueProvider.of("-SSS-of-NNN"))
+            .withSuffix(StaticValueProvider.of(null))
+            .withMinutePattern(StaticValueProvider.of(invalidMinutePattern));
+    policy.windowedFilename(1, 1, window, paneInfo, new TestOutputFileHints());
+  }
 }
```

Comment:
Please add a test to verify that when a **custom** pattern is set the **default** pattern is not being replaced.
E.g. if the directory is something like ".../recommendations/mmmm" and the `minutePattern` is `mmmm`, then the resulting directory is ".../recommendations/0047" and not ".../reco47endations/0047".

Reply:
Ah yes, I had added a test for this after our discussion yesterday. Will push it along with the other changes.